tracer/src/Datadog.Trace/Util/Http/HttpRequestUtils.cs

```diff
@@ -0,0 +1,30 @@
+// <copyright file="HttpRequestUtils.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections.Generic;
+using System.Linq;
+#if NETFRAMEWORK
+using System.Web.Routing;
+#endif
+#if !NETFRAMEWORK
+using Microsoft.AspNetCore.Routing;
+#endif
+
+namespace Datadog.Trace.Util.Http
+{
+    internal static class HttpRequestUtils
+    {
+        internal static IDictionary<string, object> ConvertRouteValueDictionary(RouteValueDictionary routeDataDict)
+        => routeDataDict.ToDictionary(
+                c => c.Key,
+                c => c.Value switch
+                    {
+                        List<RouteData> routeDataList => ConvertRouteValueList(routeDataList),
+                        _ => c.Value?.ToString()
+                    });
+
+        private static object ConvertRouteValueList(List<RouteData> routeDataList) => routeDataList.Select(x => ConvertRouteValueDictionary(x.Values)).ToList();
```

Comment:
If this is called on each web request, I would avoid using LINQ for performance reasons. Something like this avoids allocating multiple enumerators and avoids expanding the list as it grows:

```suggestion
        private static object ConvertRouteValueList(List<RouteData> routeDataList)
        {
            var list = new List<IDictionary<string, object>>(routeDataList.Count);
            
            foreach(var item in routeDataList)
            {
                list.Add(ConvertRouteValueDictionary(item.Values));
            }
            
            return list;
        }
```

Similar comment for `ToDictionary()` in line 20 above.

Reply:
In the end I removed it all in 3170ff58cb7399902a74045c473e53268c891888 
the routevalues are a dictionary, and even in case of complex objects it is turned to nested collections of keyvaluepair which is ok to pass to the encoder then to the waf